libcst/codemod/visitors/tests/test_apply_type_annotations.py

```diff
@@ -298,7 +298,7 @@ def foo(x = 1) -> Iterable[Any]:
                 """
                 from typing import Iterable, Any
 
-                def foo(x: int = 1) -> Iterable[Any]:
+                def foo(x = 1) -> Iterable[Any]:
```

Comment:
The three changes above are not really desirable from the standpoint of `MonkeyType`, which is the main client I know well: in that setting by default we don't want to overwrite existing type annotations, but we do want to guess at any types that lack existing annotation.

`pyre infer` is similar, although I think in most cases it wouldn't produce incompatible types the way MonkeyType would because the type inference algorithm works differently from sampling runtime types.

This is why I think it might be worth separating out the structural aspects of signature matching - which by default we probably do want - from the type-annotation-compatibility aspects which by default I'm guessing many clients won't want.

Reply:
To add more color to my comment

**Why the type-annotation compatibility might not be a great default**

for a function like
```
def f(x) -> Iterable[int]:
    return [x]
```
if `x` is always an int then the return type annotation is completely correct, but MonkeyType is going to infer something like
```
def f(x: int) -> list[int]: ...
```
which will fail the compatibility check and cause us to not apply the (correct and desirable) annotation of `int` for `x.`

The underlying issue is that `Iterable[int]` is a perfectly valid annotation for `list[int]` values, but libcst isn't type-system aware and therefore doesn't know this, and there's not an easy way to change that.

**Why structural compatibility probably is a good default**

On the other hand a structural mismatch is different and much less controversial, I don't think we'd ever want to partially annotate a function
```
def f(x, y):
    return 5
```
based on a stub file containing
```
def f(x: str) -> None: ...
```

Reply:
yes, that's a very good point - annotations may be compatible but not identical, and we don't have enough information about the type system here to tell that. i think i can do this with a two-pass approach - collect pairs of overlapping type annotations, let the calling code pass them to another library to determine which types are compatible, and then run the apply type annotation pass with a mapping of compatible types. it might end up being beyond the scope of this visitor if it clutters the code too much for a low-use feature, but some sort of subclass solution should be possible. for now i'll put the identical-type matching behind a flag and turn it off by default.

Reply:
btw, i did think of the false negative issue, but it seemed more important to not merge incompatible signatures, especially given that pyi files allow overloads. i guess that's a philosophical difference, i can definitely see your point about clients wanting maximal insertion of annotations by default. i haven't yet come up with a good way to handle overloads either, though i would like to at some point.